File: leetcode/82.Remove_Duplicates_from_Sorted_List_II/cpp/cpp/main.cpp
#include <vector>
struct ListNode {
    int val;
    ListNode* next;
    ListNode()
        : val(0)
        , next(nullptr)
    {
    }
    ListNode(int x)
        : val(x)
        , next(nullptr)
    {
    }
    ListNode(int x, ListNode* next)
        : val(x)
        , next(next)
    {
    }
};

class Solution {
public:
    ListNode* del_rec(ListNode* left, ListNode* right)
    {
        if (right == nullptr) {
            left->next = nullptr;
            return left;
        }
        if (left->val == right->val) {
            while (right != nullptr && right->val == left->val) {
                right = right->next;
            }
            if (right == nullptr)
                return nullptr;
            return del_rec(right, right->next);
        }

        left->next = del_rec(right, right->next);

        return left;
    }
    ListNode* deleteDuplicates(ListNode* head)
    {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        return del_rec(head, head->next);
    }
};

File: leetcode/82.Remove_Duplicates_from_Sorted_List_II/cpp/rust/src/main.rs
struct Solution;
//
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}
impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut ret = None;
        let mut tail = &mut ret;
        let mut to_delete = 200i32;

        while let Some(mut node) = head {
            head = node.next.take();
            if node.val == to_delete {
                continue;
            }
            if head.as_ref().map_or(false, |n| n.val == node.val) {
                to_delete = node.val;
                continue;
            }
            *tail = Some(node);
            tail = &mut tail.as_mut().unwrap().next;
        }

        ret
    }
}

fn main() {
    todo!();
}
